Sources/Transformer.swift

```diff
@@ -0,0 +1,177 @@
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+import AlgoliaSearch
+
+/// A Transformable data source
+public protocol Transformable {
+    associatedtype IResults
+    associatedtype IParameters
+    
+    typealias SearchResultsHandler = (_ results: IResults?, _ error: Error?) -> Void
+    
+    // Search operation
+    func search(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia params to custom backend params.
+    func map(query: Query) -> IParameters
+    
+    // Transforms the custom backend result to an Algolia result.
+    func map(results: IResults) -> SearchResults
+    
+    
+    // Disjunctive Search operation
+    func searchDisjunctiveFaceting(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia disjunctive params to custom backend params.
+    func map(query: Query, disjunctiveFacets: [String], refinements: [String : [String]]) -> IParameters
+    
+    
+    // Search for facet value operation
+    func searchForFacetValues(_ query: IParameters, searchResultsHandler: @escaping SearchResultsHandler)
+    
+    // Transforms the Algolia facet value params to custom backend params.
+    func map(query: Query?, facetName: String, matching text: String) -> IParameters
+}
+
+/// Base class used to implement a custom backend with Algolia
+/// You need to make sure to implement the following method:
+/// - search(query:searchResultsHandler:)
+/// - map(query:)
+/// - map(results:)
+open class SearchTransformer<Parameters, Results>: Transformable, Searchable {
+    
+    public typealias IParameters = Parameters
+    public typealias IResults = Results
+    
+    public init() {}
+    
+    // Transformable protocol
+    
+    open func search(_ query: Parameters, searchResultsHandler: @escaping SearchResultsHandler) {
+        fatalError("make sure to override search(query:searchResultsHandler:) for custom backend")
+    }
+    
+    open func map(query: Query) -> Parameters {
+        fatalError("make sure to override map(query:) for custom backend")
```

Comment:
From [what I've read on `fatalError`](https://cocoacasts.com/what-is-fatalerror-in-swift-and-when-to-use-it), it stops the application flow.

It's worth mentioning that on the JS client solution, only the `search()` method is required (for a custom backend implementation, it can be used to fetch data from an endpoint). Both mappers are by default **identity functions** (they return what there were given). This in order to allow mapping override on the backend solution.

I'd therefore suggest not to throw errors here, but rather to return the query as is (if we want to align our behaviors).

Reply:
That's a very good point. Unfortunately we have a swift restriction on using [default types to generics](https://stackoverflow.com/a/42015159/905425). i.e there is no way today for me to specify that the generic `Parameters` is by default equal to the concrete `Query` class. So I can't return query in the function.

One way to solve this could be to specify `Any` instead of the generic `Parameters`, but we would lose strong typing which is not "Swifty" and can lead to bugs. WDYT?

Reply:
I see. I don't know Swift very well so I won't be really relevant here.

Can't your generic `Parameters` somehow extend `Query` (as well as `Request`) to solve this problem?

Another way of solving this would be in the parent function (the function that calls the mappers). Skip the `map()` calls if the mappers are not implemented.

Reply:
Unfortunately, the 2 things you mentioned are a limitation of Swift if I want to do it in the same `SearchTransformer` class. But there is a workaround actually: I can create `DefaultSearchTransformer` that defines Parameters=Query and Results=SearchResults, something like that, WDYT?

```swift
/// Base class used to implement a custom backend with Algolia
/// You need to make sure to implement the following method:
/// - search(query:searchResultsHandler:)
open class DefaultSearchTransformer: SearchTransformer<Query, SearchResults> {
    open override func search(_ query: Query, searchResultsHandler: @escaping SearchResultsHandler) {
        fatalError("make sure to override search(query:searchResultsHandler:) for custom backend")
    }
    
    open override func map(query: Query) -> Query {
        return query
    }
    
    open override func map(results: SearchResults) -> SearchResults {
        return results
    }
}
```

Reply:
As long as it makes the implementation possible and simple for you, got for it 🙂